au/code/au/utility/factoring.hh

```diff
@@ -75,29 +76,28 @@ constexpr std::uintmax_t find_pollard_rho_factor(std::uintmax_t n) {
 
 template <typename T = void>
 struct FirstPrimesImpl {
-    static constexpr uint16_t values[] = {
+    static constexpr std::array<uint16_t, 100u> values = {
         2,   3,   5,   7,   11,  13,  17,  19,  23,  29,  31,  37,  41,  43,  47,  53,  59,
         61,  67,  71,  73,  79,  83,  89,  97,  101, 103, 107, 109, 113, 127, 131, 137, 139,
         149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233,
         239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337,
         347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439,
         443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541};
-    static constexpr std::size_t N = sizeof(values) / sizeof(values[0]);
 };
 template <typename T>
-constexpr uint16_t FirstPrimesImpl<T>::values[];
-template <typename T>
-constexpr std::size_t FirstPrimesImpl<T>::N;
+constexpr std::array<uint16_t, 100u> FirstPrimesImpl<T>::values;
 using FirstPrimes = FirstPrimesImpl<>;
 
 // Find the smallest factor which divides n.
 //
 // Undefined unless (n > 1).
 constexpr std::uintmax_t find_prime_factor(std::uintmax_t n) {
-    const auto &first_primes = FirstPrimes::values;
-
     // First, do trial division against the first N primes.
-    for (const auto &p : first_primes) {
```

Comment:
Q: Which compiler didn't support C-Style arrays? It seems OK for clang and gcc: [godbolt](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsQAJnL0WTUKgCkBgEJXr5FY1RECQ6s0wBhdPQCubJkMtcg8AGQImbAA5fwAjbFIQAGZyAAd0JWJXJm8/AKC0jKyhcMiYtnjElMdsZ2yRIhZSIlz/QINgmrqhBqaiUui4hOSHRubW/I7RvoHyyuSASgd0X1JUTi4rJIjUPxwAagskz1QlIkJhI9wLLQBBLZ297EPj0/P6Alirm/uDbaZdr4DkdPK4zqRsOxvncfuDfM59gAxAikM6qUgEDhKQ4Adlsd32hP2ZxYLlQ%2BwwTDO2HUqVIxPOIBAvgiRB4ADYAPpEfYANwkvmwSgsAFZbCKACIvKUWPE/IkK/ZGIkpIki8hEnEawk8Hjaniq/Y8LU6gCc2r%2BFvNhKSeptJv20j18sVhOkhukDpFhpF1v27Lt/odOMDOMNOL9AA5DZG/aaHTwtM6Ca6jVpVYmTYnrbqMwYs7aNQaCznpDmRXaeCKs%2ByM%2BytS7XcaMxGi5HK6aOxn40XTdaDLqNQZLUr80ODP2kqq/ubG4qDE6hxWlyaDLWh%2Bz%2ByGhzjV%2B2h9Gh6bVUktCbbXbbSk5wrbefC/sp%2BfPRqkmXX96P9akvXX2G/9%2Bh6PrGr49jeRLSEmGqQdaTqwQO0EPu6qrurB7rQe%2BjrVtBAY4Shv7gW6raOpGJplnaZbWiK6YatRVEIfsIojiKTqyhKRz4vcOLsUknE/JS1K0vSZyYEyLLCGwLDqNy%2BwsJgmCcnSmJChACy4pxCoCTyLC%2BCQhxrnQqJEIpGJYtKSIomiplCky/J%2BEKHE/HOIliayknSTyKxEKkunmVojkwimhKCPSEBabJunoPsSkcPsID7IZZwmcpShqbKGmpl5Pk8nYRxSjF2ABbcCpsU5QX7BCRCrEw%2BxZbpRWlYFtysvskkRKp6nOYyaBeS8nggrJ8nJViHUggNYCbCKnhMBNRUKpV1X7LgAAaACSAAqnIiAonieLgIgiA13E/FwSz0NwIr8IEXA6OQ6DcP1Nh2MSKxrM8WwGPwRDaKdSxIJCOCJKp5AANYgNRAB0b7SCKka%2BuysOmvWKTnVw0j8Gw4PBNdt33Vw/BKCAwTfTdp3kHAsAoCQPiUNQTTAITXAALSoP9qAg8KSQSgAVOQTP0PQvJsB8sSoLzTNsFgCRMAQABe2DixCcksLEDDEAAnuLbAEEoqBM74KhM6y0sSPzEQgywwDYHzqCpKkGAQsALI4B8kRKEzLC8ugBCYCzQhnEzZBMxCNBM5gpIsBL2AVAkSh8xIRAJJHvhMKQPj0Pz6DoKkSj8PTSiqKY2D0CISDoAA7jj/AYGwqQMAkUTsBs6iRuyTPstIFImGYRoihDWj9/w2CECQGKiUYgjCGIEicDIcjCMoaiaKT5D6EYkusBg%2BDEGQAASLBKNY2DMLckUAPKpMww/cHw5AbywW8j2QF/MDfGNCA/6ANBcb%2BjD/%2BN/wiIXSIJcy6V1/tTegq1RK7FMMAFwmB1ZEFICnEG5BkHHx9kTcg5dSAsFSG/M6F0ro/TutwWBZgmYIPVvsCAyDUFqRbm3DuXc4G937v3Wh29R76SSMqbwtd670g%2BgsL6P0Fh/QBsMYGqN0Z3yxuQKuZCAGE2JuIpYYNIbQ1hvDRGyNjDcCSCQleeMxGkyWBTBA8AIAoBrnXRgFAqBhXQIIhxaBu7ACrFoYIdB6CJ1RNQWIpDVasFIOrX%2BISmjqzPrEXQtQSa3xrhwYQZ8mD0HCSvHAsRfDAE8BIegjNb44EkmYSQmSUTxIILyGyK8aS1F0hsW%2BxtUa3RFngsJ3gcAQNMr/appBYgZGwBKaOcDXagHMQIEwDMABqBBsDlxftdW%2Bk9RDiEkHPFZi8NCkLXsYOBaBco2GMJ8QmkAljZxcP7bgTMADq%2BT9i3JpMglgDyRJ5UsE9GwMh%2BDoD6WPGyZyHDF0qW4CAHgJiBD1GECIgwY6JD1OkTIlycg%2BDaFIQoSLsizCGPCoFThkW9HGKi/IeougErGP0GFcxhikopRC9FJJmjYrhVIJYShXrrEMEQrgl1FGkLxvsJh7dO4UOAOwgeWguFP2EX8Awoj0HqNBiAaQkYIbRhxFoYcP46Jw3ZAYtGGNwZJAhiqnETokginZB3JGpE9VKNMQ4bBJMdAWMQDYkAdUiC02ca4huTduBCpYaK8Vg8QjSqwRPeQ091myE2SobZK9dm4PwYQ/VvL7XcDPrpbKtUaCCtbsK1hPcqwcMlRAAR9iEi8LlWYl1ki5LSO5XIzGPADD90jJGJM7J2x8LXJ29UGaVFOvEeTN1tiXGVscdQOxQiQDABxOqXx/jCYQCCSvSJYSIkRCiTEuJzhf5JOYEQVJ6TSFZJyXkgWhT%2BDFNGRsW6hAIR1GqYzW6dTUANN/s00hbSomdPvV9HpvB%2BB9IGSoYZJTgBjJHTQKZShZnzMWb/FZ0bZ6xvkFs5et19B6g8Qcz59gRanOBhc7IjNbn3MeeoZ5rzzjvMOdYb5d0/k%2BwBcDMl2R3BMC8MSyFIRuPMvmAiooyL6XCcxSUKlOKGXAu6EwQlLReMyfxfUClgmaXTCJXkPjjLKVlGkzwNlHLOByrTcY3G3B83MM7sAVA5IcQQxFFKneMq%2BHyudb9cg/0G1AybYat8JqFyWqRh0fMtYUiDoJsO8xGjlWmghgjFiMh0xOnTBOfVRi%2BUmO4LWzzqNPpZYswAjzEjyB9MyG4aQQA%3D%3D).

`std::array` is generally recommended per the [cppcoreguidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#slcon1-prefer-using-stl-array-or-vector-instead-of-a-c-array). But I don't like that the length has to be specified in C++14. It doesn't in C++17 😄

Reply:
Good question!  I slacked you the (Aurora-internal) build log.  I think it was `NvccCompile` choking on the particular way that we were using the array.  It was complaining about an incomplete type, but I didn't want to mess with it and decided to just try using `std::array` instead. :slightly_smiling_face: